hushline/premium.py

```diff
@@ -0,0 +1,609 @@
+            session.clear()
+            return redirect(url_for("login"))
+
+        if user.stripe_subscription_id:
+            try:
+                stripe.Subscription.modify(user.stripe_subscription_id, cancel_at_period_end=False)
+                user.stripe_subscription_cancel_at_period_end = False
+                db.session.add(user)
+                db.session.commit()
+
+                current_app.logger.info(
+                    f"Autorenew enabled for subscription {user.stripe_subscription_id} for user {user.id}"  # noqa: E501
+                )
+
+                flash("Autorenew has been enabled.")
+                return jsonify(success=True)
+            except stripe._error.StripeError as e:
+                current_app.logger.error(f"Stripe error: {e}")
+                return jsonify(success=False), 400
+
+        return jsonify(success=False), 400
+
+    @bp.route("/cancel", methods=["POST"])
+    @authentication_required
+    def cancel() -> Response | str | Tuple[Response | str, int]:
+        user = db.session.get(User, session.get("user_id"))
+        if not user:
+            session.clear()
+            return redirect(url_for("login"))
+
+        if user.stripe_subscription_id:
+            try:
+                # Cancel the subscription
+                stripe.Subscription.delete(user.stripe_subscription_id)
+
+                # Downgrade the user (the subscription ID will get removed in the webhook)
+                user.tier_id = FREE_TIER
+                db.session.add(user)
+                db.session.commit()
+
+                current_app.logger.info(
+                    f"Subscription {user.stripe_subscription_id} canceled for user {user.id}"
+                )
+
+                flash("💔 Sorry to see you go!")
+                return jsonify(success=True)
+            except stripe._error.StripeError as e:
+                current_app.logger.error(f"Stripe error: {e}")
+                return jsonify(success=False), 400
+
+        return jsonify(success=False), 400
+
+    @bp.route("/status.json", methods=["GET"])
+    @authentication_required
+    def statis() -> Response | str:
+        user = db.session.get(User, session.get("user_id"))
+        if not user:
+            session.clear()
+            return redirect(url_for("login"))
+
+        if user.tier_id == BUSINESS_TIER:
+            flash("🔥 Congratulations, you've upgraded your account!")
+
+        return jsonify({"tier_id": user.tier_id})
+
+    @bp.route("/webhook", methods=["POST"])
+    def webhook() -> Response | str | Tuple[Response | str, int]:
+        payload = request.data
+        sig_header = request.headers["STRIPE_SIGNATURE"]
+
+        # Parse the event
+        try:
+            event = stripe.Webhook.construct_event(
+                payload, sig_header, current_app.config.get("STRIPE_WEBHOOK_SECRET")
+            )
+        except ValueError as e:
+            current_app.logger.error(f"Invalid payload: {e}")
```

Comment:
Slight possibility of logging sensitive data depending on what the `ValueError` contains. Not sure there's a way to log this. Maybe the webhook has an ID we can log and then reference it out of band?

Reply:
I'm not too concerned because the webhook events shouldn't actually contain sensitive data. No credit card numbers or anything.

Reply:
I suspected no either but wanted to flag it at least